Fixes #36563 - Remove orphaned content unit cleanup from index_content

sjha4 opened the pull request:
#### What are the changes introduced in this pull request?
Move remove content unit orphan deletion to async task to avoid increasing indexing times which impacts both syncs and publishes.
#### Considerations taken when implementing this change?

#### What are the testing steps for this pull request?
1. Sync a repo. You should see the task: "Remove orphaned content units" run after completion of sync or publish CV actions asynchronously.

parthaa commented on app/models/katello/concerns/pulp_database_unit.rb:
```diff
@@ -169,7 +169,7 @@ def with_identifiers(ids)
-          where.not(:id => repository_association_class.where(:repository_id => ::Katello::Repository.all).select(unit_id_field))
+          where.not(:id => repository_association_class.select(unit_id_field))
         else
           where.not(:repository_id => ::Katello::Repository.all)
```
```suggestion
          where.not(:repository_id => ::Katello::Repository.select(:id))
```

parthaa reviewed (approved):
Seems to work ok . Checked sync and saw the new action

parthaa reviewed (approved):
Tested the remove orphan calls. Behavior looked good. ACK

ianballou commented on app/lib/actions/katello/repository/index_content.rb:
```diff
@@ -31,6 +31,10 @@ def run
         end
+
+        def finalize
+          ForemanTasks.async_task(OrphanCleanup::RemoveOrphanedContentUnits, {repo_id: input[:id]})
```
I'm questioning now if we even need to destroy orphaned units after syncing.

For the orphaned filter rule problem, the solution is this: https://github.com/Katello/katello/blob/05f9de2ae38093f5e30e96eed21882671bab6ea2/app/services/katello/content_unit_indexer.rb#L155

Otherwise, I don't think orphaned content units should cause any other problems.

Plus, if a user syncs 20 repositories, they're going to get 20 orphan cleanup tasks running, which could eat up resources on their system.

Perhaps we should just rely on the weekly orphan cleanup?

If we do that, I just suggest that we change the following error messages to tell the users to run orphan cleanup instead of run a complete sync:

- https://github.com/Katello/katello/blob/master/app/services/katello/pulp3/repository.rb#L500
- https://github.com/Katello/katello/blob/master/app/services/katello/pulp3/repository.rb#L512
- https://github.com/Katello/katello/blob/master/app/services/katello/pulp3/repository/docker.rb#L73

ianballou replied:
Actually, maybe we should change those error messages with your PR here regardless. Complete sync no longer deletes the orphaned units itself.

ianballou commented on app/lib/actions/katello/orphan_cleanup/remove_orphaned_content_units.rb:
```diff
@@ -0,0 +1,32 @@
+        end
+
+        def run
+          content_types_to_index = []
```
`content_types_to_index` makes it sound a bit like the types will be indexed instead of checked for orphans.

ianballou commented on app/lib/actions/katello/orphan_cleanup/remove_orphaned_content_units.rb:
```diff
@@ -0,0 +1,32 @@
+            repo = ::Katello::Repository.find(input[:id])
+            content_types_to_index = repo.repository_type.content_types_to_index
+          else
+            fail "Pass either a repository to determine content type or destroy_all to destroy all orphaned content units"
```
What do you think about removing the `destroy_all` param?  That way only `repo_id` would be needed and the logic would be a bit simplified. Orphan cleanup should only be deleting useless data, so I don't think a safety net is needed.

sjha4 commented:
Updated the PR to run orphan content unit deletion only in orphan cleanup task which should run once a week in deployed environments.
Removed inputs from task to make it run on all content units during the orphan cleanup.
Updated error messages to point to the rake task.

jeremylenz commented:
PR title needs an update now ;)

sjha4 commented:
Should be ready for re-review.

ianballou commented on app/services/katello/pulp3/repository.rb:
```diff
@@ -497,7 +497,8 @@ def add_content(content_unit_href, remove_all_units = false)
           raise ::Katello::Errors::Pulp3Error, "Content units that do not exist in Pulp were requested to be copied."\
-            " Please run a complete sync on the following repository: #{repository_reference.root_repository.name}. Original error: #{e.message}"
+             " Please run `foreman-rake katello:delete_orphaned_content` to fix the following repository:"\
+            " #{::Katello::Pulp3::RepositoryReference.find_by(repository_href: repository_reference.repository_href).root_repository.name}. Original error: #{e.message}"
```
```suggestion
            " #{repository_reference.root_repository.name}. Original error: #{e.message}"
```

ianballou replied:
The `repository_reference` can be used here since the method is adding content to the service class's own repo.

ianballou commented on app/lib/actions/katello/orphan_cleanup/remove_orphans.rb:
```diff
@@ -7,27 +7,16 @@ class RemoveOrphans < Actions::Base
         def plan(proxy)
           sequence do
+            if proxy.pulp_primary?
+              ::Katello::RootRepository.orphaned.destroy_all
```
Any reason why this was pulled out of the run phase?

sjha4 replied:
Not really..This was the only operation in the run phase after refactoring the content unit deletion..and this doesn't seem like something that needs to be in the run phase..

jeremylenz replied:
If there's a DB operation, should try not to put it in the run phase because `run` doesn't run in a transaction :)

ianballou replied:
Would a heavy DB operation in the planning phase eat up a Puma worker? Also we shouldn't need transactions here at least since we're deleting orphans.  But even if we did need one, couldn't we do a `transaction` block?

jeremylenz replied:
> I'm a little confused because we stick heavy DB ops in tasks in other places

Usually they're in `plan` or `finalize` though, right? Those are fine because they run in transactions by default.

sjha4 replied:
Ack to that..We tend to put only things that need API calls to other products in run phase generally. For this task, I'd be ok keeping this in run phase over the plan phase. Any failures on this are set to rescue=skip and also we do want to still plan other pulp orphan deletion actions even if the orphan deletion in katello fails. A plan phase failure wouldn't plan those pulp tasks. 

We could put this in finalize if we want. Let me know if we feel strongly about that and I can update this.

jeremylenz replied:
Makes sense to me. I trust your judgment to choose the phase correctly :)

ianballou reviewed (approved):
Works for me!
